django/apps/registry.py

```diff
@@ -344,6 +346,79 @@ def clear_cache(self):
             for model in self.get_models(include_auto_created=True):
                 model._meta._expire_cache()
 
+    @staticmethod
+    def _model_key(model_or_name):
+        """
+        Returns an (app_label, model_name) tuple from a model or string.
+        """
+        if isinstance(model_or_name, six.string_types):
+            app_label, model_name = model_or_name.split(".")
+        else:
+            # It's actually a model class.
+            app_label = model_or_name._meta.app_label
+            model_name = model_or_name._meta.object_name
+        return app_label, model_name
+
+    def lazy_model_operation(self, function, *models_or_names, **kwargs):
+        """
+        Take a function and a number of models or model names, and when all
+        the corresponding models have been loaded, call the function with
+        the model classes and any optional as its arguments.
+
+        Each element of models_or names may be
+            * an actual model class
+            * a string of the form "app_label.ModelName"
+            * a tuple of strings of the form (app_label, model_name)
+
+        The function passed to this method must accept n positional arguments,
+        where n=len(models_or_names), plus whatever keyword arguments you pass
+        in kwargs.
+        """
+
+        if not self._class_prepared_handler_registered:
+            from django.db.models import signals
+            signals.class_prepared.connect(self.do_pending_lookups)
+            self._class_prepared_handler_registered = True
+
+        # Eagerly parse all model strings so we can fail immediately
+        # if any are plainly invalid.
+        model_keys = [self._model_key(m) if not isinstance(m, tuple) else m
+                      for m in models_or_names]
+
+        # If this function depends on more than one model, recursively call add
+        # for each, partially applying the given function on each iteration.
+        model_key, more_models = model_keys[0], model_keys[1:]
+        if more_models:
+            inner_function = function
+            function = lambda model, **kwargs: (
```

Comment:
This is clever, but it's really hard to parse; I think it may be too clever. Why not just allow the keys in `_pending_lookups` to be lists of model keys instead? Sure, that would require a bit more code when handling `_pending_lookups`, but I think it might be clearer overall.

Or you could simplify the patch by dropping the capability for `lazy_model_operation` to accept multiple models - does Django itself actually need that capability anywhere? If not, then I definitely think it's a case of YAGNI and should be done in a follow-up patch, if ever.

(I am going to argue in a later comment that `lazy_model_operation` should not be public API, so my approach to this patch is that it should be a refactoring of Django internals for simplicity and clarity, not an attempt to meet any other theoretical needs of a public API.)